contrib/win32/win32compat/misc.c

```diff
@@ -237,9 +242,33 @@ dlclose(HMODULE handle)
 FARPROC 
 dlsym(HMODULE handle, const char *symbol)
 {
-	return GetProcAddress(handle, symbol);
+	void *ptr = GetProcAddress(handle, symbol);
+	if (ptr == NULL)
+		last_dlerror = GetLastError();
+	return ptr;
 }
 
+char *
+dlerror()
+{
+	static char *message = NULL;
+	if (message != NULL) {
+		LocalFree(message);
+		message = NULL;
+	}
+	if (last_dlerror == ERROR_SUCCESS)
+		return NULL;
+
+	DWORD length = FormatMessageA(
+		FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+		NULL, last_dlerror, 0, (char *) &message, 0, NULL);
+	last_dlerror = ERROR_SUCCESS;
+
+	if (length == 0)
+		return "Failed to format error message";
+
+	return message;
```

Comment:
What's the expectation on how "message" is freed? Looks like a simple free() on this can lead to issues as its supposed to be LocalFree()d

If this is true, you will have to return and xcopy version and LocalFree original version.

Reply:
Per the [spec](http://pubs.opengroup.org/onlinepubs/9699919799/): The returned pointer might be invalidated or the string content might be overwritten by a subsequent call to dlerror() in the same thread (if dlerror() is thread-safe) or in any thread (if dlerror() is not thread-safe). The returned pointer might also be invalidated if the calling thread is terminated.

In other words, the user does not free the message. This is why there is a LocalFree() at the top of the function.